compiler/optimizer/abstractinterpreter/OMRIDTBuilder.cpp

```diff
@@ -0,0 +1,324 @@
+                                                NULL,
+                                                rootMethod->containingClass(),
+                                                0,
+                                                0,
+                                                rootMethod,
+                                                _rootSymbol,
+                                                _rootSymbol->getMethodKind() == TR::MethodSymbol::Kinds::Virtual || _rootSymbol->getMethodKind() == TR::MethodSymbol::Kinds::Interface ,
+                                                _rootSymbol->getMethodKind() == TR::MethodSymbol::Kinds::Interface,
+                                                bcInfo,
+                                                comp());
+
+   TR_CallTarget *rootCallTarget = new (region()) TR_CallTarget(
+                                    rootCallSite,
+                                    _rootSymbol,
+                                    rootMethod,
+                                    NULL,
+                                    rootMethod->containingClass(),
+                                    NULL);
+
+   //Initialize IDT
+   _idt = new (region()) TR::IDT(region(), rootCallTarget, _rootSymbol, _rootBudget, comp());
+
+   TR::IDTNode* root = _idt->getRoot();
+
+   //generate the CFG for root call target
+   TR::CFG* cfg = self()->generateControlFlowGraph(rootCallTarget);
+
+   if (!cfg) //Fail to generate a CFG
+      return _idt;
+
+   //add the IDT decendants
+   buildIDT2(root, NULL, -1, _rootBudget, NULL);
+
+   if (traceBIIDTGen)
+      traceMsg(comp(), "\n+ IDTBuilder: Finish building TR::IDT |\n");
+
+   return _idt;
+   }
+
+void OMR::IDTBuilder::buildIDT2(TR::IDTNode* node, TR::vector<TR::AbsValue*, TR::Region&>* arguments, int32_t callerIndex, int32_t budget, TR_CallStack* callStack)
+   {
+   TR::ResolvedMethodSymbol* symbol = node->getResolvedMethodSymbol();
+   TR_ResolvedMethod* method = node->getResolvedMethod();
+
+   TR_CallStack* nextCallStack = new (region()) TR_CallStack(comp(), symbol, method, callStack, budget, true);
+
+   // Check if current method has been interpreted.
+   // If so, there is no need to run the abstract interpretation on this method since they have the same static properties.
+   TR::IDTNode* interpretedMethodIDTNode = checkIfMethodIsInterpreted(method);
+
+   if (interpretedMethodIDTNode)
+      {
+      // since they are the same method, they should have the same inlining method summary.
+      node->setInliningMethodSummary(interpretedMethodIDTNode->getInliningMethodSummary());
+      uint32_t staticBenefit = computeStaticBenefit(node->getInliningMethodSummary(), arguments);
+
+      node->setStaticBenefit(staticBenefit);
+
+      // Same methods should have same IDT decendents.
+      // IDT is built in DFS order, at this point, we have all the descendants so it is safe to copy all the descendants.
+      copyDescendants(interpretedMethodIDTNode, node);
```

Comment:
OK, looking at everything including `performAbstractInterpretation()` (at least, the implementation in eclipse-openj9/openj9#17813), it looks like the arguments are ignored when searching for call site targets, as you said. That does seem like quite a missed opportunity though. If abstract interpretation has determined a type for the receiver of a call, then that type really should feed into call target selection somehow. Not that that should block merging

However, this code currently assumes that the downstream `performAbstractInterpretation()` won't use the arguments to inform the selection of call targets - which perhaps it shouldn't assume - and there's still the issue of differing budgets. So I don't think I'm sold on copying the subtree

Reply:
Yes, the `copyDescendants` indeed has some problems with budgets and I'm currently working on a solution for that. I will try to submit it before tomorrow.

Reply:
```
buildIDT() --> buildIDT2() --> generateControlFlowGraph() --> performAbstractInterpretation() --> addNodesToIDT() 
             |                                                                                              |  
             |______________________________________________________________________________________________| 
```
Above is the process of building the IDT. When performing abstract interpretation, it will not only generate a ininling method summary but also upodates `arguments`. The inlining method summary includes all potential optimization and their constraints, and then we will use the updated arguments to test each constraints in the inlining method summary to see if a potential optimization can be unlocked, aka `calcuteStaticBenefit()`.
For the problem of `copyDescendants()`, it indeed will miss some node, if the budget of `fromNode` is less than `toNode`. There is a way to copy those missed node, which is continue run the abstract interpretation and even if no remaining budget but don't add them into the IDT, and then store those nodes in a hashmap for `copyDescendants()`. However, the other problem is the static benefit of copied nodes may vary depends on their diffrent arguments. To solve this problem, we need to recalculate static benefit which requires the arguments, and those arguments are gotten from abstract interpretation. Therefore, although we can copy descendants instead of performing abstract interpretation for a quicker inlining, but it seems impossible to get the updated arguments and recalculate static benefit without performing abstract interpretation.
For a correct implementation, we decided remove all code related to copyDescendants for now and performing abstract interpretation for all nodes. We will run benchmark agian later to see how this will affect the performance of this inliner.

Reply:
> we decided remove all code related to copyDescendants for now

OK, thanks, that makes perfect sense to me